src/puppetlabs/trapperkeeper/services/status/ringutils.clj

```diff
@@ -2,47 +2,84 @@
 ;; probably be moved into a shared library or some such.
 (ns puppetlabs.trapperkeeper.services.status.ringutils
   (:require [clojure.tools.logging :as log]
+            [cheshire.core :as json]
+            [ring.util.response :as response]
+            [schema.core :as schema]
             [slingshot.slingshot :refer [try+]]
             [puppetlabs.kitchensink.core :as ks]))
 
 ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;;; Schemas
+
+(def ResponseType (schema/enum :json :plain))
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
 ;;; Public
+(defmacro with-json-response [status & body]
+  `(-> (do ~@body)
+       json/encode
+       response/response
+       (response/status ~status)
+       (response/content-type "application/json; charset=utf-8")))
+
+(defmacro with-plain-response [status & body]
+  `(-> (do ~@body)
+       response/response
+       (response/status ~status)
+       (response/content-type "text/plain; charset=utf-8")))
 
-(defn wrap-request-data-errors
+
+(schema/defn ^:always-validate wrap-request-data-errors
   "A ring middleware that catches slingshot errors of :type
   :request-dava-invalid and returns a 400."
-  [handler]
-  (fn [request]
-    (try+ (handler request)
-          (catch #(contains? #{:request-data-invalid :service-status-version-not-found}
-                             (:type %)) e
-            {:status 400
-             :body e}))))
-
-(defn wrap-schema-errors
+  [handler type :- ResponseType]
+  (let [code 400
+        response (fn [e]
+                   (case type
+                     :json (with-json-response code e)
+                     :plain (with-plain-response code (:message e))))]
+    (fn [request]
+      (try+ (handler request)
+            (catch #(contains? #{:request-data-invalid :service-status-version-not-found}
+                               (:type %)) e
```

Comment:
The binding of the caught exception to `e` is pretty easy to miss here. I think it's always bad style to have a new argument on the same line as the ending of a multi-line argument form; if you've got a multi-line argument then it should be the only thing on its lines.